flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java

```diff
@@ -501,6 +501,22 @@ private <P extends ResponseBody> CompletableFuture<P> submitRequest(
                     }
                 });
 
+        // [FLINK-32583] If connectFuture failed instantly but channelFuture is unresolved, it may
+        // mean the executor service Netty is using has shut down, in which case the above listener
+        // to complete channelFuture will never run
+        if (connectFuture.isDone() && !connectFuture.isSuccess() && !channelFuture.isDone()) {
```

Comment:
Should we add a comment why we're not handling the success case? Essentially, we're trying to work around a bug in the netty code, aren't we?
```
    /**
     * Adds the specified listener to this future.  The
     * specified listener is notified when this future is
     * {@linkplain #isDone() done}.  If this future is already
     * completed, the specified listener is notified immediately.
     */
```
The JavaDoc of `Future.addListener` states that the listener would be informed if the future is already completed (which includes the successful and the exceptional case). But that doesn't match the implementation, apparently.

Right now, we're missing the `isCancelled` and `isSuccess` case handling. It's not clear whether these cases can actually happen. But since we're (at least) covering the success case in the listener implementation above, wouldn't it be reasonable to cover it here as well?

Reply:
While trying to get a better picture of the various cases we might need to handle to get full coverage here I took a step back and tried to get a broader view of this.

I verified that the event loop/executor for both the connection as well as its listeners is the same, the NioEventLoopGroup `bootstrap` is configured with elsewhere in RestClient, and I saw that when the client is shut down, this group is asked to shut down gracefully with a "quiet period" of 0 and a default timeout of 10 seconds. As far as I can tell, already-scheduled tasks are allowed to run to completion.

I believe we could cover the vast majority of cases by simply checking `isRunning` at the top of `submitRequest`—if the client has already been closed and graceful shutdown of the group initiated, we should immediately reject any new requests.

The challenge in particular is covering the case where the client is shut down after the call to `bootstrap.connect()` but before the listeners are notified. Depending on configured timeouts, the connect task may run to completion or time out, but not be able to notify its listeners and thus resolve `channelFuture`. I don't think we have any way to detect this within `submitRequest` itself.

The only solution that comes to mind here would be to:
- Add each `channelFuture` created in `submitRequest` to a class-level collection
- In the listener attached to `connectFuture`, remove the `channelFuture` from that collection after calling `complete`/`completeExceptionally`
- In the listener in `shutdownInternally` that is notified when the event loop shuts down, resolve all remaining futures in the collection exceptionally

Having written it down it doesn't sound as complex as I first thought, but it's still clearly beyond what I had originally hoped for this change.

@XComp Thoughts? Should I go with simply checking whether the client has already been closed at the top of `submitRequest`, implement this "future cleanup" logic, or both?

Reply:
The diff for RestClient if we just have it preemptively check `isRunning` (the test can remain the same):

```diff
diff --git a/flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java b/flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java
index 2055a2227cf..e9f57c2f4ce 100644
--- a/flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java
+++ b/flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java
@@ -488,18 +488,22 @@ public class RestClient implements AutoCloseableAsync {
 
     private <P extends ResponseBody> CompletableFuture<P> submitRequest(
             String targetAddress, int targetPort, Request httpRequest, JavaType responseType) {
-        final ChannelFuture connectFuture = bootstrap.connect(targetAddress, targetPort);
-
         final CompletableFuture<Channel> channelFuture = new CompletableFuture<>();
 
-        connectFuture.addListener(
-                (ChannelFuture future) -> {
-                    if (future.isSuccess()) {
-                        channelFuture.complete(future.channel());
-                    } else {
-                        channelFuture.completeExceptionally(future.cause());
-                    }
-                });
+        if (isRunning.get()) {
+            final ChannelFuture connectFuture = bootstrap.connect(targetAddress, targetPort);
+
+            connectFuture.addListener(
+                    (ChannelFuture future) -> {
+                        if (future.isSuccess()) {
+                            channelFuture.complete(future.channel());
+                        } else {
+                            channelFuture.completeExceptionally(future.cause());
+                        }
+                    });
+        } else {
+            channelFuture.completeExceptionally(new IOException("RestClient is closed"));
+        }
 
         return channelFuture
                 .thenComposeAsync(
```

Reply:
@XComp After manually trying out various orderings of the threaded operations, I've come to the conclusion that as long as `addListener` is called before the client is closed, the listener will be run. This means there is only a minuscule period between the non-blocking call to `bootstrap.connect()` and `connectFuture.addListener()` where this situation could arise.

So again, I think the vast likelihood of hitting this issue is resolved with the simple `isRunning` check from the diff in my previous comment, while 100% coverage would require tracking the futures and resolving them on client shutdown, though with a higher chance of introducing another bug. Its behavior would also be difficult to test.

Including the future tracking would look something like this:

```diff
diff --git a/flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java b/flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java
index 2055a2227cf..c986f7c6292 100644
--- a/flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java
+++ b/flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java
@@ -96,8 +96,10 @@ import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.IdentityHashMap;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 import java.util.ServiceLoader;
 import java.util.concurrent.CompletableFuture;
@@ -127,6 +129,9 @@ public class RestClient implements AutoCloseableAsync {
 
     private final String urlPrefix;
 
+    private final Map<CompletableFuture<Channel>, Void> responseChannelFutures =
+            Collections.synchronizedMap(new IdentityHashMap<>());
+
     @VisibleForTesting List<OutboundChannelHandlerFactory> outboundChannelHandlerFactories;
 
     public RestClient(Configuration configuration, Executor executor)
@@ -252,6 +257,8 @@ public class RestClient implements AutoCloseableAsync {
                             .shutdownGracefully(0L, timeout.toMilliseconds(), TimeUnit.MILLISECONDS)
                             .addListener(
                                     finished -> {
+                                        notifyResponseFuturesOfShutdown();
+
                                         if (finished.isSuccess()) {
                                             terminationFuture.complete(null);
                                         } else {
@@ -265,6 +272,16 @@ public class RestClient implements AutoCloseableAsync {
         return terminationFuture;
     }
 
+    private void notifyResponseFuturesOfShutdown() {
+        responseChannelFutures
+                .keySet()
+                .forEach(
+                        future ->
+                                future.completeExceptionally(
+                                        new IOException("RestClient is closed")));
+        responseChannelFutures.clear();
+    }
+
     public <
                     M extends MessageHeaders<EmptyRequestBody, P, EmptyMessageParameters>,
                     P extends ResponseBody>
@@ -488,18 +505,25 @@ public class RestClient implements AutoCloseableAsync {
 
     private <P extends ResponseBody> CompletableFuture<P> submitRequest(
             String targetAddress, int targetPort, Request httpRequest, JavaType responseType) {
-        final ChannelFuture connectFuture = bootstrap.connect(targetAddress, targetPort);
-
         final CompletableFuture<Channel> channelFuture = new CompletableFuture<>();
+        responseChannelFutures.put(channelFuture, null);
 
-        connectFuture.addListener(
-                (ChannelFuture future) -> {
-                    if (future.isSuccess()) {
-                        channelFuture.complete(future.channel());
-                    } else {
-                        channelFuture.completeExceptionally(future.cause());
-                    }
-                });
+        if (isRunning.get()) {
+            final ChannelFuture connectFuture = bootstrap.connect(targetAddress, targetPort);
+
+            connectFuture.addListener(
+                    (ChannelFuture future) -> {
+                        responseChannelFutures.remove(channelFuture);
+
+                        if (future.isSuccess()) {
+                            channelFuture.complete(future.channel());
+                        } else {
+                            channelFuture.completeExceptionally(future.cause());
+                        }
+                    });
+        } else {
+            channelFuture.completeExceptionally(new IOException("RestClient is closed"));
+        }
 
         return channelFuture
                 .thenComposeAsync(
```

Reply:
I guess your last proposal makes sense. Could you add this as a proper commit to make it reviewable? And it's not too complicated.